### PR TITLE
Native token refactoring

### LINKED ISSUE #1431: Add Native Token Feature
Add Native Token Feature and remove the native tokens fields from the outputs. Also update the rent cost if https://github.com/iotaledger/iota-sdk/pull/1065 already merged then

### PR DESCRIPTION
**Pull request by thibault-martinez**
Fixes #1431

Changes:
- Native token is now a feature
- Only Basic and Foundry can hold a NativeTokenFeature
- Only one native token per output

### CODE REVIEW

File: bindings/nodejs/lib/types/wallet/address.ts [hunk 1]
```diff
@@ -35,12 +35,12 @@ export interface SendParams {
...
-/** Address with native tokens */
-export interface SendNativeTokensParams {
+/** Address with native token */
+export interface SendNativeTokenParams {
     /** The Bech32 address. */
     address: Bech32Address;
-    /** The Native Tokens to send. */
-    nativeTokens: [TokenId, u256][];
+    /** The Native Token to send. */
+    nativeToken: [TokenId, u256][];
```
Comment (DaughterOfMars) [inline bindings/nodejs/lib/types/wallet/address.ts]: still an array?

File: bindings/python/iota_sdk/types/send_params.py [hunk 2]
```diff
@@ -32,17 +32,17 @@ class SendParams():
...
-        native_tokens: The native tokens to send.
-        return_address: The address to return the native tokens to if not claimed.
-        expiration: The expiration timestamp until native tokens can be claimed.
+        native_tokens: The native token to send.
+        return_address: The address to return the native token to if not claimed.
+        expiration: The expiration timestamp until the native token can be claimed.
     """
     address: str
-    native_tokens: List[NativeToken]
+    native_tokens: NativeToken
```
Comment (DaughterOfMars) [inline bindings/python/iota_sdk/types/send_params.py]: singular

File: sdk/src/types/block/output/foundry.rs [hunk 3]
```diff
@@ -265,8 +255,7 @@ impl FoundryOutputBuilder {
         verify_allowed_features(&immutable_features, FoundryOutput::ALLOWED_IMMUTABLE_FEATURES)?;
 
         let mut output = FoundryOutput {
-            amount: 0,
-            native_tokens: NativeTokens::from_set(self.native_tokens)?,
+            amount: 1u64,
```
Comment (DaughterOfMars) [inline sdk/src/types/block/output/foundry.rs]: Why 1?

Comment (thibault-martinez) [reply sdk/src/types/block/output/foundry.rs]: Failed merge, reverted

File: bindings/python/iota_sdk/types/send_params.py (hunk 2 above)
Comment (thibault-martinez) [reply bindings/python/iota_sdk/types/send_params.py]: Done

File: bindings/nodejs/lib/types/wallet/address.ts (hunk 1 above)
Comment (thibault-martinez) [reply bindings/nodejs/lib/types/wallet/address.ts]: Done

File: bindings/nodejs/lib/types/wallet/bridge/wallet.ts [hunk 4]
```diff
@@ -9,7 +9,7 @@ import type {
 } from '../../client';
 import type {
     SendParams,
-    SendNativeTokensParams,
+    SendNativeTokenParams,
```
Comment (Thoralf-M) [inline bindings/nodejs/lib/types/wallet/bridge/wallet.ts:12]: SendNativeTokenParams still has `nativeTokens: [TokenId, u256][];`, `[]` should be removed

Comment (Thoralf-M) [reply bindings/nodejs/lib/types/wallet/bridge/wallet.ts:12]: Same in Python (List[] there)

Comment (Thoralf-M) [reply bindings/nodejs/lib/types/wallet/bridge/wallet.ts:12]: Nvm, I reviewed the last commit :trollface:

Comment (thibault-martinez) [reply bindings/nodejs/lib/types/wallet/bridge/wallet.ts:12]: Hmm it doesn't?

Comment (thibault-martinez) [reply bindings/nodejs/lib/types/wallet/bridge/wallet.ts:12]: Ah, no worries

File: sdk/src/types/block/output/feature/native_token.rs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    use serde::{Deserialize, Serialize};
+
+    use super::*;
+    use crate::types::block::{output::TokenId, Error};
+
+    #[derive(Serialize, Deserialize)]
+    struct NativeTokenFeatureDto {
+        #[serde(rename = "type")]
+        kind: u8,
+        token_id: TokenId,
```
Comment (Thoralf-M) [inline sdk/src/types/block/output/feature/native_token.rs:41]: This field is just `id` in JSON

Comment (thibault-martinez) [reply sdk/src/types/block/output/feature/native_token.rs:41]: Good catch, fixed

File: sdk/src/wallet/operations/transaction/prepare_transaction.rs [hunk 6]
```diff
@@ -7,6 +7,7 @@ use instant::Instant;
 use packable::bounded::TryIntoBoundedU16Error;
 
 #[cfg(feature = "events")]
+// TODO why is this not used anymore? Are we not triggering an event anymore?
```
Comment (Thoralf-M) [inline sdk/src/wallet/operations/transaction/prepare_transaction.rs]: No remainder address generation anymore, removed the imports in another PR
```suggestion
```

File: sdk/src/types/block/payload/signed_transaction/transaction.rs [hunk 7]
```diff
@@ -447,16 +446,6 @@ fn verify_outputs<const VERIFY: bool>(outputs: &[Output], visitor: &ProtocolPara
                 return Err(Error::InvalidTransactionAmountSum(amount_sum as u128));
             }
 
-            if let Some(native_tokens) = native_tokens {
-                native_tokens_count = native_tokens_count.checked_add(native_tokens.len() as u8).ok_or(
-                    Error::InvalidTransactionNativeTokensCount(native_tokens_count as u16 + native_tokens.len() as u16),
-                )?;
-
-                if native_tokens_count > NativeTokens::COUNT_MAX {
```
Comment (Thoralf-M) [inline sdk/src/types/block/payload/signed_transaction/transaction.rs:455]: This check is still listed in https://github.com/iotaledger/tips/blob/tip45/tips/TIP-0045/tip-0045.md
`The count of all distinct native tokens present in outputs must not be larger than Max Native Token Count.`
Really not needed anymore?

Comment (thibault-martinez) [reply sdk/src/types/block/payload/signed_transaction/transaction.rs:455]: I've commented, we definitely said we'd remove it

**Review by Thoralf-M — DISMISSED** [review]

**Review by Thoralf-M — DISMISSED** [review]

**Review by Alex6323 — APPROVED** [review]

**Review by Thoralf-M — APPROVED** [review]

**Pull request merged**